src/OrleansGoogleUtils/Providers/Streams/GooglePubSubBatchContainer.cs

```diff
@@ -0,0 +1,83 @@
+﻿using Newtonsoft.Json;
+using Orleans.Providers.Streams.Common;
+using Orleans.Runtime;
+using Orleans.Streams;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Orleans.Serialization.Providers.Streams
+{
+    public class GooglePubSubBatchContainer : IBatchContainer
+    {
+        [JsonProperty]
+        private EventSequenceTokenV2 sequenceToken;
+
+        [JsonProperty]
+        private readonly List<object> events;
+
+        [JsonProperty]
+        private readonly Dictionary<string, object> requestContext;
+
+        public Guid StreamGuid { get; }
+
+        public String StreamNamespace { get; }
+
+        public StreamSequenceToken SequenceToken => sequenceToken;
+
+        internal EventSequenceTokenV2 RealSequenceToken
+        {
+            set { sequenceToken = value; }
+        }
+
+        [JsonConstructor]
+        public GooglePubSubBatchContainer(
+            Guid streamGuid,
+            String streamNamespace,
+            List<object> events,
+            Dictionary<string, object> requestContext,
+            EventSequenceTokenV2 sequenceToken)
+            : this(streamGuid, streamNamespace, events, requestContext)
+        {
+            this.sequenceToken = sequenceToken;
+        }
+
+        public GooglePubSubBatchContainer(Guid streamGuid, String streamNamespace, List<object> events, Dictionary<string, object> requestContext)
+        {
+            StreamGuid = streamGuid;
+            StreamNamespace = streamNamespace;
+            this.events = events ?? throw new ArgumentNullException(nameof(events), "Message contains no events");
+            this.requestContext = requestContext;
+        }
+
+        public IEnumerable<Tuple<T, StreamSequenceToken>> GetEvents<T>()
+        {
+            return events.OfType<T>().Select((e, i) => Tuple.Create<T, StreamSequenceToken>(e, sequenceToken.CreateSequenceTokenForEvent(i)));
+        }
+
+        public bool ShouldDeliver(IStreamIdentity stream, object filterData, StreamFilterPredicate shouldReceiveFunc)
+        {
+            foreach (object item in events)
+            {
+                if (shouldReceiveFunc(stream, filterData, item))
+                    return true; // There is something in this batch that the consumer is intereted in, so we should send it.
+            }
+            return false; // Consumer is not interested in any of these events, so don't send.
+        }
+
+        public bool ImportRequestContext()
+        {
+            if (requestContext != null)
+            {
+                RequestContext.Import(requestContext);
+                return true;
+            }
+            return false;
+        }
+
```

Comment:
There are tons of "same things" around in Streaming internal APIs. If you look at AWS SQS, Azure Queue and this implementation, they are almost the very same code. I've already mentioned it before that we should refactor out all those common code in a Streaming V2 framework. The same for Storage and Membership...